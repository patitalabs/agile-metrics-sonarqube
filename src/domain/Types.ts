import { SaCollectorConfig, SaMetricItem } from './static-analysis/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface SaCollectorService {
  fetch(collectorConfig: SaCollectorConfig): Promise<SaMetricItem[]>;
}
