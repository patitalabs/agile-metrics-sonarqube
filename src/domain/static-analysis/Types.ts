export interface SaMetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export class SaCollectorConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;

  constructor({ projectName, since, until = null, teamName }) {
    this.projectName = projectName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
