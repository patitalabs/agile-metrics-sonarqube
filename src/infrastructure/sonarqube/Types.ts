export class ProjectMetrics {
  createdAt: Date;
  teamName: string;
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export interface SonarConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;
}

export interface SonarService {
  projectMetrics(sonarConfig: SonarConfig): Promise<ProjectMetrics[]>;
}

export interface SonarClient {
  getData(url: string): Promise<any>;
}

export interface SonarRepository {
  projectMetrics(sonarConfig: SonarConfig): Promise<ProjectMetrics[]>;
}
