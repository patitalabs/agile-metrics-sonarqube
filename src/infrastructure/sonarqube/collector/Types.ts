export interface SonarMetricItem {
  projectName: string;
  alertStatus: string;
  qualityGateDetails: string;
  bugs: number;
  newBugs: number;
  reliabilityRating: number;
  newReliabilityRating: number;
  vulnerabilities: number;
  newVulnerabilities: number;
  securityRating: number;
  newSecurityRating: number;
  codeSmells: number;
  newCodeSmells: number;
  sqaleRating: number;
  newMaintainabilityRating: number;
  sqaleIndex: number;
  newTechnicalDebt: string;
  coverage: number;
  newCoverage: number;
  newLinesToCover: number;
  tests: number;
  duplicatedLinesDensity: number;
  newDuplicatedLinesDensity: number;
  duplicatedBlocks: number;
  ncloc: number;
  nclocLanguageDistribution: string;
  newLines: number;
  version: string;
}

export class StaticAnalysisCollectorConfig {
  projectName: string;
  teamName: string;
  since: string;
  until?: string;

  constructor({ projectName, since, until = null, teamName }) {
    this.projectName = projectName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
