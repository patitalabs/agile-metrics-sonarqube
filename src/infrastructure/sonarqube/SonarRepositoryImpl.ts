import { Converters } from './Converters';
import {
  ProjectMetrics,
  SonarClient,
  SonarConfig,
  SonarRepository,
} from './Types';

export class SonarRepositoryImpl implements SonarRepository {
  constructor(private readonly sonarClient: SonarClient) {}

  async projectMetrics(sonarConfig: SonarConfig): Promise<ProjectMetrics[]> {
    // TODO pagination
    const measures = await this.sonarClient.getData(
      SonarRepositoryImpl.searchHistoryUrl(sonarConfig)
    );
    const analysisHistory = await this.sonarClient.getData(
      SonarRepositoryImpl.analysisHistoryUrl(sonarConfig)
    );

    return Converters.toProjectMetrics(measures, analysisHistory, sonarConfig);
  }

  private static analysisHistoryUrl(sonarConfig: SonarConfig) {
    return `/api/project_analyses/search?category=VERSION&project=${
      sonarConfig.projectName
    }&ps=499&from=${sonarConfig.since}${this.untilParam(sonarConfig)}`;
  }

  private static searchHistoryUrl(sonarConfig: SonarConfig): string {
    return `/api/measures/search_history?component=${
      sonarConfig.projectName
    }&metrics=alert_status,quality_gate_details,bugs,new_bugs,reliability_rating,new_reliability_rating,vulnerabilities,new_vulnerabilities,security_rating,new_security_rating,code_smells,new_code_smells,sqale_rating,new_maintainability_rating,sqale_index,new_technical_debt,coverage,new_coverage,new_lines_to_cover,tests,duplicated_lines_density,new_duplicated_lines_density,duplicated_blocks,ncloc,ncloc_language_distribution,projects,new_lines&ps=499&from=${
      sonarConfig.since
    }${this.untilParam(sonarConfig)}`;
  }

  private static untilParam(sonarConfig: SonarConfig) {
    let untilParam = '';
    if (sonarConfig.until) {
      untilParam = `&to=${sonarConfig.until}`;
    }
    return untilParam;
  }
}
