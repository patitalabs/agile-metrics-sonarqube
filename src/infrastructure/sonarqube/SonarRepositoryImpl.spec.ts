import { SonarRepositoryImpl } from './SonarRepositoryImpl';
import * as searchHistoryFakeData from './test/search-history-response.json';
import * as searchProjectFakeData from './test/search-project-analysis.json';
import { SonarClient, SonarRepository } from './Types';

describe('SonarRepositoryImpl', () => {
  const sonarClient: SonarClient = {
    getData: (url): Promise<any> => {
      let result = {};
      if (/.*api\/measures\/search_history/.test(url)) {
        result = searchHistoryFakeData;
      } else if (/.*api\/project_analyses\/search/.test(url)) {
        result = searchProjectFakeData;
      }
      return Promise.resolve(result);
    },
  };
  const sonarRepository: SonarRepository = new SonarRepositoryImpl(sonarClient);

  it('should get projectMetrics', async () => {
    const sonarConfig = {
      projectName: 'someProject',
      since: '2018-11-20',
      teamName: 'someTeam',
    };

    const data = await sonarRepository.projectMetrics(sonarConfig);
    expect(data).toMatchSnapshot();
  });
});
