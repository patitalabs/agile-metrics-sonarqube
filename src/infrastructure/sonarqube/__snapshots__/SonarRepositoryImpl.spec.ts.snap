// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SonarRepositoryImpl should get projectMetrics 1`] = `
Array [
  Object {
    "alertStatus": "OK",
    "bugs": 0,
    "codeSmells": 30,
    "coverage": 86.3,
    "createdAt": 2018-12-10T10:45:09.000Z,
    "duplicatedBlocks": 8,
    "duplicatedLinesDensity": 2.5,
    "ncloc": 4534,
    "nclocLanguageDistribution": "java=4534",
    "newBugs": 0,
    "newCodeSmells": 0,
    "newCoverage": 0,
    "newDuplicatedLinesDensity": 0,
    "newLines": 0,
    "newLinesToCover": 0,
    "newMaintainabilityRating": 0,
    "newReliabilityRating": 0,
    "newSecurityRating": 0,
    "newTechnicalDebt": "",
    "newVulnerabilities": 0,
    "projectName": "someProject",
    "qualityGateDetails": "{\\"level\\":\\"OK\\",\\"conditions\\":[],\\"ignoredConditions\\":false}",
    "reliabilityRating": 1,
    "securityRating": 1,
    "sqaleIndex": 300,
    "sqaleRating": 1,
    "teamName": "someTeam",
    "tests": 74,
    "version": undefined,
    "vulnerabilities": 0,
  },
  Object {
    "alertStatus": "OK",
    "bugs": 0,
    "codeSmells": 27,
    "coverage": 87,
    "createdAt": 2018-12-11T08:17:33.000Z,
    "duplicatedBlocks": 8,
    "duplicatedLinesDensity": 2.5,
    "ncloc": 4530,
    "nclocLanguageDistribution": "java=4530",
    "newBugs": 0,
    "newCodeSmells": 0,
    "newCoverage": 0,
    "newDuplicatedLinesDensity": 0,
    "newLines": 0,
    "newLinesToCover": 0,
    "newMaintainabilityRating": 0,
    "newReliabilityRating": 0,
    "newSecurityRating": 0,
    "newTechnicalDebt": "",
    "newVulnerabilities": 0,
    "projectName": "someProject",
    "qualityGateDetails": "{\\"level\\":\\"OK\\",\\"conditions\\":[{\\"metric\\":\\"new_security_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_reliability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_maintainability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_coverage\\",\\"op\\":\\"LT\\",\\"period\\":1,\\"error\\":\\"80\\",\\"actual\\":\\"92.3076923076923\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_duplicated_lines_density\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"3\\",\\"actual\\":\\"0.0\\",\\"level\\":\\"OK\\"}],\\"ignoredConditions\\":false}",
    "reliabilityRating": 1,
    "securityRating": 1,
    "sqaleIndex": 300,
    "sqaleRating": 1,
    "teamName": "someTeam",
    "tests": 74,
    "version": undefined,
    "vulnerabilities": 0,
  },
  Object {
    "alertStatus": "OK",
    "bugs": 0,
    "codeSmells": 29,
    "coverage": 86.9,
    "createdAt": 2018-12-13T11:58:40.000Z,
    "duplicatedBlocks": 8,
    "duplicatedLinesDensity": 2.5,
    "ncloc": 4535,
    "nclocLanguageDistribution": "java=4535",
    "newBugs": 0,
    "newCodeSmells": 0,
    "newCoverage": 0,
    "newDuplicatedLinesDensity": 0,
    "newLines": 0,
    "newLinesToCover": 0,
    "newMaintainabilityRating": 0,
    "newReliabilityRating": 0,
    "newSecurityRating": 0,
    "newTechnicalDebt": "",
    "newVulnerabilities": 0,
    "projectName": "someProject",
    "qualityGateDetails": "{\\"level\\":\\"OK\\",\\"conditions\\":[{\\"metric\\":\\"new_security_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_reliability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_maintainability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_coverage\\",\\"op\\":\\"LT\\",\\"period\\":1,\\"error\\":\\"80\\",\\"actual\\":\\"93.54838709677419\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_duplicated_lines_density\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"3\\",\\"actual\\":\\"0.0\\",\\"level\\":\\"OK\\"}],\\"ignoredConditions\\":false}",
    "reliabilityRating": 1,
    "securityRating": 1,
    "sqaleIndex": 300,
    "sqaleRating": 1,
    "teamName": "someTeam",
    "tests": 76,
    "version": undefined,
    "vulnerabilities": 0,
  },
  Object {
    "alertStatus": "OK",
    "bugs": 0,
    "codeSmells": 29,
    "coverage": 86.9,
    "createdAt": 2018-12-14T10:52:28.000Z,
    "duplicatedBlocks": 8,
    "duplicatedLinesDensity": 2.5,
    "ncloc": 4535,
    "nclocLanguageDistribution": "java=4535",
    "newBugs": 0,
    "newCodeSmells": 0,
    "newCoverage": 0,
    "newDuplicatedLinesDensity": 0,
    "newLines": 0,
    "newLinesToCover": 0,
    "newMaintainabilityRating": 0,
    "newReliabilityRating": 0,
    "newSecurityRating": 0,
    "newTechnicalDebt": "",
    "newVulnerabilities": 0,
    "projectName": "someProject",
    "qualityGateDetails": "{\\"level\\":\\"OK\\",\\"conditions\\":[{\\"metric\\":\\"new_security_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_reliability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_maintainability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_coverage\\",\\"op\\":\\"LT\\",\\"period\\":1,\\"error\\":\\"80\\",\\"actual\\":\\"93.54838709677419\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_duplicated_lines_density\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"3\\",\\"actual\\":\\"0.0\\",\\"level\\":\\"OK\\"}],\\"ignoredConditions\\":false}",
    "reliabilityRating": 1,
    "securityRating": 1,
    "sqaleIndex": 300,
    "sqaleRating": 1,
    "teamName": "someTeam",
    "tests": 76,
    "version": undefined,
    "vulnerabilities": 0,
  },
  Object {
    "alertStatus": "OK",
    "bugs": 0,
    "codeSmells": 29,
    "coverage": 86.9,
    "createdAt": 2018-12-14T11:41:48.000Z,
    "duplicatedBlocks": 8,
    "duplicatedLinesDensity": 2.5,
    "ncloc": 4535,
    "nclocLanguageDistribution": "java=4535",
    "newBugs": 0,
    "newCodeSmells": 2,
    "newCoverage": 93.5483870967742,
    "newDuplicatedLinesDensity": 0,
    "newLines": 64,
    "newLinesToCover": 19,
    "newMaintainabilityRating": 1,
    "newReliabilityRating": 1,
    "newSecurityRating": 1,
    "newTechnicalDebt": "0",
    "newVulnerabilities": 0,
    "projectName": "someProject",
    "qualityGateDetails": "{\\"level\\":\\"OK\\",\\"conditions\\":[{\\"metric\\":\\"new_security_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_reliability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_maintainability_rating\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"1\\",\\"actual\\":\\"1\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_coverage\\",\\"op\\":\\"LT\\",\\"period\\":1,\\"error\\":\\"80\\",\\"actual\\":\\"93.54838709677419\\",\\"level\\":\\"OK\\"},{\\"metric\\":\\"new_duplicated_lines_density\\",\\"op\\":\\"GT\\",\\"period\\":1,\\"error\\":\\"3\\",\\"actual\\":\\"0.0\\",\\"level\\":\\"OK\\"}],\\"ignoredConditions\\":false}",
    "reliabilityRating": 1,
    "securityRating": 1,
    "sqaleIndex": 300,
    "sqaleRating": 1,
    "teamName": "someTeam",
    "tests": 76,
    "version": "1.0",
    "vulnerabilities": 0,
  },
]
`;
